### Brief Overview 

* In this assignment, you will practice some data wrangling functionality commonly used in real-world projects.

* This assignment is designed around two dataset

* The Medicaid Data per State  

* IRS Statistics of Income (SOI) dataset


* The final product of this assignment will be a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  * Which medications account for the bulk of a state's spending   
  * Which medications are prescribed more frequently in one state than another.
  * etc.

* The following URLs provide the datasets required for this assignment:
  * [medicaid_data.csv](https://www.dropbox.com/s/u6ctjqxnk0wxpbk/medicaid_data.csv?dl=1)
  * [medicaid_enrollment.tsv](https://www.dropbox.com/s/969mohzhpu78r10/medicaid_enrollment.tsv?dl=1)
  * [tax_data.csv](https://www.dropbox.com/s/zm37nu7vnirha4m/tax_data.csv?dl=1)


 ### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame` or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitly asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external modules will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, assuming you are using Jupyter Notebooks, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora of functionality available through the `pandas` package. 


### 1. Loading and exploring the tax data

Import Pandas and numpy using their appropriate alias here


In [128]:
# WRITE YOUR CODE HERE

import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (`tax_data.csv`) into a `DataFrame` called `tax_data`. This file can be downloaded from the `URL` provided above.

* This file `tax_data.csv` was preprocessed from the original file I obtained at the following URL:
  * [15zpallagi.csv](https://www.irs.gov/pub/irs-soi/15zpallagi.csv}

In [129]:
## WRITE YOUR CODE HERE

tax_data = pd.read_csv('tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [130]:
## WRITE YOUR CODE HERE

tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0          1    AL        0         1  836320.0  481570.0  109790.0  233260.0   
1          1    AL        0         2  494830.0  206630.0  146250.0  129390.0   
2          1    AL        0         3  261250.0   80720.0  139280.0   36130.0   
3          1    AL        0         4  166690.0   28510.0  124650.0   10630.0   
4          1    AL        0         5  212660.0   19520.0  184320.0    4830.0   
5          1    AL        0         6   55360.0    2950.0   49260.0     350.0   
6          1    AL    35004         1    1490.0     970.0     230.0     280.0   
7          1    AL    35004         2    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 131 columns]

*  Modify `tax_data` to uppercase all the header names. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can use `tax_data.columns`, which returns a `Series` of the column names.

* The resulting column name should look as follows:

```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

* Standardizing column names is a good way to avoid having to guess whether the column header is in lowercase or uppercase.

In [131]:
## WRITE YOUR CODE HERE

tax_data = tax_data.rename(columns=lambda x: x.upper())
print(tax_data.columns)

Index(['STATEFIPS', 'STATE', 'ZIPCODE', 'AGI_STUB', 'N1', 'MARS1', 'MARS2',
       'MARS4', 'PREP', 'N2',
       ...
       'N10300', 'A10300', 'N85530', 'A85530', 'N85300', 'A85300', 'N11901',
       'A11901', 'N11902', 'A11902'],
      dtype='object', length=131)


* What is the total number of entries (also called observations or instances) in `tax_data` dataset?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [132]:
## WRITE YOUR CODE HERE

print(tax_data.size)

21837438


* If `STATEFIPS` is the header label of the first column in `tax_data`, write code to print the title of the 32nd column.  
  * You should use a single python expression
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [133]:
## WRITE YOUR CODE HERE

print(tax_data.columns[32])

A00900


* If `STATEFIPS` is the first column label (index 0), what is the index of the column labeled `N10300`?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [134]:
## WRITE YOUR CODE HERE

print(tax_data.columns.get_loc("N10300"))

121


- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown here)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zip codes in TX, 9,238 in NY, etc.


In [135]:
## WRITE YOUR CODE HERE

#Get Series of States
unique_zipByState = tax_data.groupby("STATE")["ZIPCODE"].nunique()
print(unique_zipByState.sort_values(ascending=False))
unique_zipByState = unique_zipByState.to_frame().reset_index()
print(unique_zipByState)

STATE
TX    1619
NY    1540
CA    1485
PA    1368
IL    1231
OH     998
FL     920
MI     892
MO     890
IA     827
VA     793
MN     792
NC     725
WI     714
IN     674
GA     667
KY     656
KS     598
TN     590
AL     577
OK     547
NJ     547
WV     512
WA     497
AR     490
NE     486
MA     482
LA     453
MD     405
CO     395
SC     376
ME     370
MS     370
OR     356
AZ     295
ND     287
SD     287
CT     264
VT     238
NH     232
MT     225
ID     213
NM     209
UT     186
NV     128
WY     109
RI      71
HI      60
DE      57
AK      56
DC      24
Name: ZIPCODE, dtype: int64
   STATE  ZIPCODE
0     AK       56
1     AL      577
2     AR      490
3     AZ      295
4     CA     1485
5     CO      395
6     CT      264
7     DC       24
8     DE       57
9     FL      920
10    GA      667
11    HI       60
12    IA      827
13    ID      213
14    IL     1231
15    IN      674
16    KS      598
17    KY      656
18    LA      453
19    MA      482
20    MD      405
21    ME 

* Identify the position of `HI` in the list of zip code counts per state generated in the previous question
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [136]:
## WRITE YOUR CODE HERE
print("HI position: ", unique_zipByState['STATE'].loc[lambda x: x=="HI"].index.values[0])


HI position:  11


### 2. Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [137]:
## WRITE YOUR CODE HERE

nb_invalid_zip = int(tax_data.loc[tax_data.ZIPCODE == 0, 'ZIPCODE'].count())
print(nb_invalid_zip)

306


* Run the line in the code cell below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
* If the code cell below returns an error, then change your answer above so that the value returned is a number.

In [138]:
assert(type(nb_invalid_zip) == int)

* Remove from `tax_data` all the instances where the zip code is `0` and save the resulting `DataFrame` to a variable named `tax_data_valid_zip`

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [139]:
## WRITE YOUR CODE HERE

tax_data_valid_zip = tax_data[tax_data.ZIPCODE != 0]
print(tax_data_valid_zip.head(5))

    STATEFIPS STATE  ZIPCODE  AGI_STUB      N1  MARS1  MARS2  MARS4   PREP  \
6           1    AL    35004         1  1490.0  970.0  230.0  280.0  700.0   
7           1    AL    35004         2  1350.0  630.0  360.0  300.0  610.0   
8           1    AL    35004         3   970.0  310.0  490.0  140.0  450.0   
9           1    AL    35004         4   620.0  110.0  470.0   30.0  300.0   
10          1    AL    35004         5   620.0   40.0  560.0   20.0  320.0   

        N2  ...  N10300   A10300  N85530  A85530  N85300  A85300  N11901  \
6   2160.0  ...   690.0    610.0     0.0     0.0     0.0     0.0   120.0   
7   2540.0  ...  1140.0   3019.0     0.0     0.0     0.0     0.0   210.0   
8   2160.0  ...   930.0   5009.0     0.0     0.0     0.0     0.0   200.0   
9   1610.0  ...   620.0   5190.0     0.0     0.0     0.0     0.0   150.0   
10  1770.0  ...   620.0  10129.0     0.0     0.0     0.0     0.0   220.0   

    A11901  N11902  A11902  
6     94.0  1290.0  2792.0  
7    301.0  1130

* Run the line below to confirm that the operation worked as expected

  * The assertion below verifies that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [140]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [141]:
## WRITE YOUR CODE HERE 

#Row Count that contains NaN
print(len(tax_data_valid_zip) - len(tax_data_valid_zip.dropna()))

#All NaN Count
nb_missing_values = tax_data_valid_zip.isna().sum().sum()
print(nb_missing_values)

0
0


* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values (`NaN`)

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [142]:
## WRITE YOUR CODE HERE

tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()
print(tax_data_valid_zip_cleaned)

        STATEFIPS STATE  ZIPCODE  AGI_STUB      N1   MARS1   MARS2   MARS4  \
6               1    AL    35004         1  1490.0   970.0   230.0   280.0   
7               1    AL    35004         2  1350.0   630.0   360.0   300.0   
8               1    AL    35004         3   970.0   310.0   490.0   140.0   
9               1    AL    35004         4   620.0   110.0   470.0    30.0   
10              1    AL    35004         5   620.0    40.0   560.0    20.0   
11              1    AL    35004         6    60.0     0.0    50.0     0.0   
12              1    AL    35005         1  1350.0   750.0   190.0   390.0   
13              1    AL    35005         2   980.0   370.0   230.0   350.0   
14              1    AL    35005         3   490.0   150.0   210.0   120.0   
15              1    AL    35005         4   250.0    50.0   180.0    40.0   
16              1    AL    35005         5   190.0     0.0   160.0     0.0   
17              1    AL    35005         6     0.0     0.0     0

* Run the line below to confirm that the operation worked as expected. The assertion below tests that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [143]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code
* The function `compute_percentile_zipcode` below computes the percentile income per zip code
* By default `percentile=0.5`, i.e., the function computes the median
* Read the code and make sure you understand what it does before moving on to the next question

In [144]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zip code in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * You can use any of the submethods discussed in the Split-Apply-Combine paradigm.
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`

* The resulting `Series` should look like the following ( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [145]:
## WRITE YOUR CODE HERE

#print(tax_data_valid_zip_cleaned.head(5))
print(compute_percentile_zip(tax_data_valid_zip_cleaned, percentile=0.65))

#computed row
tax_data_valid_zip_cleaned['computed_test'] = tax_data_valid_zip_cleaned.apply(lambda row : (compute_percentile_zip(row, percentile=0.65)))

#computation test
print(tax_data_valid_zip_cleaned['computed_test'])

#sorted order Series
zip_rev_all = tax_data_valid_zip_cleaned.sorted_order(ascending=False)


36.647147147147145


KeyError: ('N1', 'occurred at index STATEFIPS')

* What are the three zip codes with the most significant 65th percentile value for income?
 * I.e., the top three zip codes in your sorted listed

In [146]:
## WRITE YOUR CODE HERE

print(zip_rev_all['ZIPCODE'].head(3))

NameError: name 'zip_rev_all' is not defined

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file can be downloaded from the URL provided above. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (e.g., less than 6-8 GB)


In [147]:
## WRITE YOUR CODE HERE

medicaid_data = pd.read_csv('medicaid_data.csv')

* Modify `medicaid_data` to capitalize all column names 

  * If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [148]:
## WRITE YOUR CODE HERE

medicaid_data = medicaid_data.rename(columns=lambda x: x.upper())

* Familiarize yourself with the data
  * the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  * The remaining column names are self-explanatory
* Explore the number of lines and columns in the data
* Check that your column headers are capitalized

In [149]:
## WRITE YOUR CODE HERE

print("Size:", medicaid_data.size)
print("Shape:", medicaid_data.shape)
print(medicaid_data.columns)


Size: 16955460
Shape: (1695546, 10)
Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')


* If you explore the `LOCATION` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical.
* Are there any states that have more than one value for `LOCATION`. 
  * This question can be answered by using aggregation and the Split-Apply-Combine paradigm
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [150]:
## WRITE YOUR CODE HERE

#HI Location Phenomena - LOCATION values are all same
db = medicaid_data[medicaid_data["STATE"] == "HI"]
print(db["LOCATION"].head(5))

#Checking for the same phenomena for other states
count_LOCATION = medicaid_data.groupby(['STATE', 'LOCATION']).size().reset_index(name='count')
print(count_LOCATION.head())

154    (21.1098, -157.5311)
294    (21.1098, -157.5311)
318    (21.1098, -157.5311)
397    (21.1098, -157.5311)
557    (21.1098, -157.5311)
Name: LOCATION, dtype: object
  STATE              LOCATION  count
0    AK   (61.385, -152.2683)  12265
1    AL    (32.799, -86.8073)  23547
2    AR   (34.9513, -92.3809)  18276
3    AZ  (33.7712, -111.3877)  27348
4    CA    (36.17, -119.7462)  76996


Yes. Based by the aggregation analysis, there are states that have identical LOCATION values as indicated by the 'count' column

* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state.  That is, we need to compare per capita. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.

 
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescriptions in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).


https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained [here](https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html)

* A parsed/processed version (`medicaid_enrollment.tsv`) can be downloaded from the URL provided above. Use `pandas` to load the `medicaid_enrollment.tsv` file into DataFrame named `medicaid_enrollment`

In [151]:
## WRITE YOUR CODE HERE

medicaid_enrollment= pd.read_csv("medicaid_enrollment.tsv", sep='\t')

* Modify `medicaid_enrollment` `DataFrame` to capitalize all the column names 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by manually uppercasing the columns

In [152]:
## WRITE YOUR CODE HERE

medicaid_enrollment = medicaid_enrollment.rename(columns=lambda x: x.upper())
print(medicaid_enrollment.columns)

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')


* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`
* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing value means in this context

In [153]:
## WRITE YOUR CODE HERE

#Notice: NaN can be either NaN or 'n/a'. Hence, we must take account all possible 'missing value' data
print(medicaid_enrollment.loc[medicaid_enrollment["STATE"] == "Guam"])
medicaid_enrollment = medicaid_enrollment.dropna()

medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[medicaid_enrollment_cleaned['STATE'] != 'Guam']
print(medicaid_enrollment_cleaned)

   STATE TOTAL MEDICAID ENROLLEES
12  Guam                      n/a
                   STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky        

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization



In [154]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"] = medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"].apply(lambda x: int(x.replace(",", "")))
print(medicaid_enrollment_cleaned.head(5))

        STATE  TOTAL MEDICAID ENROLLEES
0     Alabama                   1050989
1      Alaska                    164783
3     Arizona                   1740520
4    Arkansas                    762166
5  California                  13096861


* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and make sure that the data type is `int`

In [155]:
## WRITE YOUR CODE HERE

type(medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"])
#dtype property
medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"].dtype

dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

* We can use the shared State information across both tables to associate both tables (SQL JOIN).

* However, `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name

  * We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

* Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.

  * A brief description of what the code does is included in the comments
  
* Note that the code below may throw an error if `lxml` is not installed on your machine. You should install any Python packages that Python complains about before proceeding.

In [156]:
import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US STATE',	'POSTAL ABBREVIATION', 	'STANDARD ABBREVIATION'] to ['US STATE', 'ABBREVIATION', `STD ABBREVIATION`]
  * You can hard code this operation

In [157]:
## WRITE YOUR CODE HERE

Codes_abbreviations = Codes_abbreviations.rename({"POSTAL ABBREVIATION": "ABBREVIATION", "STANDARD ABBREVIATION": "STD ABBREVIATION"}, axis=1)
Codes_abbreviations.head(5)

US STATE ABBREVIATION STD ABBREVIATION
0     Alabama           AL             Ala.
1      Alaska           AK           Alaska
2     Arizona           AZ            Ariz.
3    Arkansas           AR             Ark.
4  California           CA           Calif.

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 

* Save the results to a variable named `medicaid_enrollment_cleaned_with_zip`

* `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online (e.g., `https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`) and in a notebook under `Data Wrangling - Advanced` module. See for instance:


* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [158]:
## WRITE YOUR CODE HERE

#medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how="inner", on="US STATE")
#print(medicaid_enrollment_cleaned_with_zip)

dfinal = medicaid_enrollment_cleaned.merge(medicaid_enrollment_cleaned, Codes_abbreviations, on="STATE", how = 'inner')

TypeError: merge() got multiple values for argument 'how'

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`

  * Remove the column and make sure the data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):



```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....

```

In [ ]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned = medicaid_enrollment_cleaned.drop(columns=['STATE'],  axis=1)
print(medicaid_enrollment_cleaned.head(5))

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not relevant. This answer uses the same approach as the one used to `merge` the tables above. 

In [ ]:
## WRITE YOUR CODE HERE 


- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [ ]:
## WRITE YOUR CODE HERE 



- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [ ]:
## WRITE YOUR CODE HERE 



* Write a single Pandas expression to print all the lines containing `NDC` `61958180101` in the Pennsylvania ("PA")

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [ ]:
## WRITE YOUR CODE HERE 


#### Computing the normalized `UNITS REIMBURSED` per drug 



* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.



- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612



```python

total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 

```



- And the numebr of Medicaid Enrollees in "PA" is 2,569,232





```python

total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]

```

- Therefore, the UNITS REIMBURSED per enrollee for "HARVONI" is  0.00413041718303





```python

print(total_amount_reimbursed/total_enrollees_PA)

```



- Rather than work directly with the ratios, we are going to compute and work with their logarithm (np.log2) instead.



  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures



- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`





- Your final result should be a `Series` that looks like the following ( `...` represents data that is not shown here):
  * Note that negative values are the result of the `log` transformation 

```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...
AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...
MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626
...
```



In [ ]:
## WRITE YOUR CODE HERE 



- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`
- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - `index` is the two-letter state symbol 
  - column names are the `NDC` codes 
- The `DataFrame` should be formatted as in the image below
  - Hint: simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [ ]:
## WRITE YOUR CODE HERE 


#### Exploring the Data 

* What is the drug with the highest log-normalized ratio in Hawaii?


In [159]:
## WRITE YOUR CODE HERE 


* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)

In [160]:
## WRITE YOUR CODE HERE 

* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii? (Not graded)

In [161]:
## WRITE YOUR ANSWER HERE 

* Find and list all unique `NDC`s for which the difference between the largest and second large log-scale ratio-by-state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ratio of  `-12.623428`

In [162]:
## WRITE YOUR CODE HERE 


* The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states
   * Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   * The second highest state has a log value of -12.623428 (0.000158478197834722)
* Oklahoma is not a high-incidence state for cancer
* Could the following explain what is happening in Oklahoma?
  * The following is clinical trial conducted by the University of Oklahoma

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10